File: level3/단어변환.py
from collections import deque

def solution(begin, target, words):
    if target not in words:
        return 0
    
    visit = {}
    candidate = [[] for _ in range(len(begin))]
    for word in words:
        visit[word] = False
        for i, w in enumerate(word):
            candidate[i].append(w)
    candidate = [list(set(c)) for c in candidate]
    
    q = deque()
    q.append([0,begin])
    
    answer = 0
    while q:
        n, w = q.popleft()
        
        if w == target:
            answer = n
            break
            
        for i, can in enumerate(candidate):
            for c in can:
                dw = w[:i]+c+w[i+1:]
                if dw in visit.keys() and not visit[dw]:
                    visit[dw] = True
                    q.append([n+1,dw])
    return answer

File: level3/단어변환.cpp
#include <string>
#include <vector>
#include <queue>
#include <iostream>
using namespace std;

struct Word{
    int cnt;
    string word;
};

int solution(string begin, string target, vector<string> words) {
    int w_len = begin.size(), list_size = words.size();
    
    queue<Word> q;
    q.push({0,begin});
    
    int answer = 0;
    vector<bool> visit(list_size,false);
    while(!q.empty()){
        int cnt = q.front().cnt;
        string word = q.front().word;
        q.pop();
        
        if(word == target){
            answer = cnt;
            break;
        }
        for(int i = 0; i < list_size; i++){
            if(visit[i]){
                continue;
            }
            
            int diff = 0;
            for(int j = 0; j < w_len; j++){
                if(words[i].at(j) != word.at(j)){
                    diff++;
                }
                if(diff > 1){
                    break;
                }
            }
            if(diff <= 1){
                visit[i] = true;
                q.push({cnt+1,words[i]});
            }
        }
        
    }
    
    return answer;
}
